### PR TITLE
Added support for activity server side retry

### PR DESCRIPTION
**Pull request by mfateev**
The change is API compatible as ActivityOptions already had RetryOptions. 
The previous implementation used the client side retry, the new switched to the server side.
The changes include addition of the server side retry to the TestWorkflowService.

### CODE REVIEW

File: src/main/java/com/uber/cadence/internal/testservice/TestWorkflowMutableStateImpl.java [hunk 1]
```diff
@@ -1075,24 +1077,62 @@ public void failActivityTask(String activityId, RespondActivityTaskFailedRequest
...
           activity.action(StateMachines.Action.FAIL, ctx, request, 0);
-          activities.remove(activityId);
-          ctx.unlockTimer();
-          scheduleDecision(ctx);
+          if (isTerminalState(activity.getState())) {
+            activities.remove(activityId);
+            ctx.unlockTimer();
+            scheduleDecision(ctx);
+          } else {
+            addActivityRetryTimer(ctx, activity);
```
Comment (meiliang86) [inline src/main/java/com/uber/cadence/internal/testservice/TestWorkflowMutableStateImpl.java]: Time skipping seems not working inside retry window.

Comment (mfateev) [reply src/main/java/com/uber/cadence/internal/testservice/TestWorkflowMutableStateImpl.java]: Fixed

File: src/main/java/com/uber/cadence/activity/Activity.java [hunk 2]
```diff
@@ -238,6 +239,34 @@ public static void heartbeat(Object details) throws CancellationException {
...
+   * the server would attempt to dispatch another activity task to retry according to the retry
+   * options. If there was heartbeat details reported by the activity from the failed attempt, the
+   * details would be delivered along with the activity task for the retry attempt. The activity
+   * could extract the details by {@link #getHeartbeatDetails(Class, Object)}() and resume from the
+   * progress.
+   *
+   * @param detailsClass type of the heartbeat details
+   * @param defaultValue value to return if no heartbeat was ever emitted
+   */
+  public static <V> V getHeartbeatDetails(Class<V> detailsClass, V defaultValue) {
```
[the remark this answers is not in the document]
Comment (mfateev) [reply src/main/java/com/uber/cadence/activity/Activity.java]: Good idea. Changed to optional.

File: src/main/java/com/uber/cadence/internal/sync/ActivityExecutionContextImpl.java [hunk 3]
```diff
@@ -101,6 +102,25 @@ public void recordActivityHeartbeat(Object details) throws ActivityCompletionExc
     }
   }
 
+  @Override
+  public <V> V getHeartbeatDetails(Class<V> detailsClass, Type detailsType, V defaultValue) {
+    lock.lock();
+    try {
+      if (lastDetails != null) {
+        @SuppressWarnings("unchecked")
+        V ld = (V) this.lastDetails;
```
[the remark this answers is not in the document]
Comment (mfateev) [reply src/main/java/com/uber/cadence/internal/sync/ActivityExecutionContextImpl.java]: Good catch. Fixed. Updated unit test to execute this path.

File: src/main/java/com/uber/cadence/internal/testservice/TestWorkflowMutableStateImpl.java (hunk 1 above)
Comment (mfateev) [reply src/main/java/com/uber/cadence/internal/testservice/TestWorkflowMutableStateImpl.java]: Fixed

File: src/main/java/com/uber/cadence/activity/Activity.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mfateev) [reply src/main/java/com/uber/cadence/activity/Activity.java]: Done

File: src/main/java/com/uber/cadence/internal/sync/ActivityExecutionContextImpl.java [hunk 4]
```diff
@@ -58,6 +59,7 @@
   private boolean doNotCompleteOnReturn;
   private final long heartbeatIntervalMillis;
   private Optional<Object> lastDetails;
+  private boolean heartbeatCalled;
```
Comment (meiliang86) [inline src/main/java/com/uber/cadence/internal/sync/ActivityExecutionContextImpl.java]: Call it hasOutstandingHeartbeat?

**Pull request merged**